pages/admin/settings.js

```diff
@@ -0,0 +1,129 @@
+import React, { Component } from 'react'
+import { connect } from 'unistore/react'
+
+import Router from '../../lib/router'
+import { actions } from '../../lib/store'
+import { isAdmin } from '../../lib/utils/roles'
+import NotLoggedIn from '../../components/NotLoggedIn'
+import AdminHeader from '../../components/admin/AdminHeader'
+import { LoadingState } from '../../components/common/LoadingState'
+import useForm from 'react-hook-form'
+
+function OSMesaSettings (props) {
+  const { register, handleSubmit } = useForm()
+  const onSubmit = props.handleSubmit
+
+  return (
+    <form className='form' onSubmit={handleSubmit(onSubmit)}>
+      <div className='form__input-unit'>
+        <label className='form__label' htmlFor='setting-osmesa-db'>OSMesa DB Connection String</label>
+        <input type='text' name='osmesa-db' defaultValue={props.settings['osmesa-db']} ref={register} />
+      </div>
+      <div className='form__input-unit'>
+        <label className='form__label' htmlFor='setting-osmesa-s3-prefix'>OSMesa Tile S3 Prefix</label>
+        <input type='text' name='osmesa-s3-prefix' defaultValue={props.settings['osmesa-s3-prefix']} ref={register} />
+      </div>
+      <div className='form__input-unit'>
+        <label className='form__label' htmlFor='setting-osmesa-api-key'>OSMesa S3 AWS API key</label>
+        <input type='text' name='osmesa-api-key' defaultValue={props.settings['osmesa-api-key']} ref={register} />
+      </div>
+      <div className='form__input-unit'>
+        <label className='form__label' htmlFor='setting-osmesa-db'>OSMesa S3 AWS API secret</label>
+        <input type='text' name='osmesa-api-secret' defaultValue={props.settings['osmesa-api-secret']} ref={register} />
+      </div>
+
+      <input type='submit' className='button' value='submit' />
+    </form>
+  )
+}
+
+export class AdminSettings extends Component {
+  constructor () {
+    super()
+
+    this.state = {
+      loading: true,
+      settings: {
+        'osmesa-db': '',
+        'osmesa-s3-prefix': '',
+        'osmesa-api-key': '',
+        'osmesa-api-secret': ''
+      }
+    }
+    this.handleSubmit = this.handleSubmit.bind(this)
+  }
+
+  async componentDidMount () {
+    try {
+      await this.props.getAuthenticatedUser()
+      await this.props.getSettings()
+      this.setState({ loading: false })
+    } catch (err) {
+      console.error(err)
+      this.setState({ loading: false })
+    }
+  }
+
+  async handleSubmit (data) {
```

Comment:
Consider encrypting settings before saving to database, especially because AWS keys are involved. There is a Js crypto api that could be used https://caniuse.com/#search=crypto . The settings would of course need to be decrypted in memory to use the key+secret. `handleSubmit` may not be the place to do that, but I wanted to log it somewhere.

Reply:
This has been handled by the DB settings model encrypting and decrypting on storage.